processing/stenosis_connector.go

```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		return nil, err
+	}
+	sConn.Client = api.NewStenosisServiceQueryClient(conn)
+
+	go func() {
+		defer conn.Close()
+
+		for flow := range sConn.FlowNotifyChan {
+			var myAlerts []types.Entry
+			aval, exist := sConn.Cache.Get(flow.FlowID)
+			if exist {
+				log.Debugf("flow with existing alert finished: %v", flow)
+				myAlerts = aval.([]types.Entry)
+				outParsed, err := sConn.submit(&flow)
+				if err != nil {
+					log.Error(err)
+					for _, a := range myAlerts {
+						forwardChan <- []byte(a.JSONLine)
+					}
+				} else {
+					if forwardChan != nil {
+						for _, a := range myAlerts {
+							var ev types.EveOutEvent
+							// annotate alerts with flows and forward
+							err := json.Unmarshal([]byte(a.JSONLine), &ev)
+							if err != nil {
+								log.Error(err)
+							}
+							if ev.ExtraInfo == nil {
+								ev.ExtraInfo = &types.ExtraInfo{
+									StenosisInfo: outParsed,
+								}
+							} else {
+								ev.ExtraInfo.StenosisInfo = outParsed
+							}
+							var jsonCopy []byte
+							jsonCopy, _ = json.Marshal(ev)
+							forwardChan <- jsonCopy
+						}
+					}
+				}
+				sConn.Cache.Delete(flow.FlowID)
+			}
+		}
+	}()
+	return sConn, nil
+}
+
+func (s *StenosisConnector) handleError(err error) error {
+	s.ErrorCount++
+	if s.ErrorCount == maxErrorCount {
+		log.Warning("maximum reported error count exceeded, disabling output")
+		return nil
+	} else if s.ErrorCount > maxErrorCount {
+		return nil
+	} else {
+		return err
+	}
+}
+
+func (s *StenosisConnector) resetErrors() {
+	if s.ErrorCount > maxErrorCount {
+		log.Warning("re-enabling error reporting")
+	}
+	s.ErrorCount = 0
+}
+
+// Accept registers the given Entry into the connector's cache setup.
+func (s *StenosisConnector) Accept(e *types.Entry) {
+	var myAlerts []types.Entry
+	aval, exist := s.Cache.Get(e.FlowID)
+	if exist {
+		myAlerts = aval.([]types.Entry)
+	} else {
+		myAlerts = make([]types.Entry, 0)
+	}
+	myAlerts = append(myAlerts, *e)
+	s.Cache.Set(e.FlowID, myAlerts, cache.DefaultExpiration)
+}
+
+func (s *StenosisConnector) submit(e *types.Entry) (interface{}, error) {
+	var ev types.EveOutEvent
+	if err := json.Unmarshal([]byte(e.JSONLine), &ev); err != nil {
+		return nil, err
+	}
+	log.Debug(string(e.JSONLine))
+
+	query := &task.Query{
+		Type: task.QueryType_FLOW_PARAM,
+		Content: &task.Query_FlowParam{FlowParam: &task.FlowParam{
+			Network:     strings.ToLower(ev.Proto),
+			SrcHostPort: fmt.Sprintf("%s:%d", ev.SrcIP, ev.SrcPort),
+			DstHostPort: fmt.Sprintf("%s:%d", ev.DestIP, ev.DestPort),
+		}},
+	}
+	if ev.Flow != nil && ev.Flow.Start != nil {
+		query.AfterTime, _ = ptypes.TimestampProto(ev.Flow.Start.Time.Add(-s.TimeBracket))
+	}
+	if ev.Flow != nil && ev.Flow.End != nil {
+		query.BeforeTime, _ = ptypes.TimestampProto(ev.Flow.End.Time.Add(s.TimeBracket))
+	}
+	// prepare context
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, s.Timeout)
+	defer cancel()
+
+	nowTime := time.Now()
```

Comment:
Warning: `time.Now().UTC()` is probably what is wanted?

Reply:
This does not matter, we are just using this to calculate a relative time difference:
```Go
nowTime := time.Now()
...
log.Debugf("query took %v", time.Since(nowTime))
```